[Declan Lee Ler] iP

declanleeler opened the pull request:
# Duke
> "_We do not remember days. We remember moments._" - Cesare Pavese [(source)](https://www.brainyquote.com/quotes/cesare_pavese_109425)

Duke remembers all, pass everything to Duke. It's
- text based
- easy to learn and use
- ~fast~ VERY VERY FAST

All you need to do is 
1. download it from [here](https://github.com/declanleeler/ip)
2. double click it
3. add your tasks
4. let it manage your tasks for you ⛷️⛷️⛷️ 

And it is **FREE!!!!!!!**
Features:
- [x] Managing Tasks
- [ ] Managing Deadlines
- [ ] Reminders
--------
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

declanleeler committed:
Duke class: skeletal version that can interact with user

Duke class only prints out DUKE output and does not take in input.

The class now greets the user, echoes the input commands, and exits upon "bye".

declanleeler committed:
Add functionality to store user input

Storing the input allows us to print all the items entered back so that the user is reminded of what they need. Using an arraylist is beneficial compared to an array as we are unsure of how long the users input will be.

declanleeler committed:
Implement feature to mark status of tasks.

The list offers no tracking ability of the items. Doing so would allow us to enhance user experience.

let's create a new class to store the information of a task (e.g. details and status).

declanleeler committed:
Add support for tracking 3 types of tasks

The 3 types of tasks have common behaviors and so inherit from the Task.
- Deadline
- Event
- ToDo

Let's create more detailed levels of tasks to allow users more options when using Duke.

declanleeler committed:
Test using the I/O redirection technique

Let's automate testing to confirm there are no regressions and to observe changes in the program behavior more easily.

declanleeler committed:
Implement custom exception for error handling

Let's handle the following possible user error:
- unknown command given
- missing date for events/deadlines
- missing description for events/deadlines/todos

declanleeler committed:
Remove redundant code and add exception case

New case of error handling where the user marks/unmarks a non-existent task number

declanleeler committed:
Add support for deleting tasks from the list

User can now delete tasks from the list using the "delete" command. Error handling implemented should they delete a non-existent task.

declanleeler committed:
Add function for I/O operations to save tasks

Let's allow users to store the list of tasks so they can access it without using the app

declanleeler committed:
Add understanding of dates for tasks

Tasks with dates such as deadlines and events will be required to follow a strict format of yyyy-mm-dd.

declanleeler committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add function for I/O operations to save tasks

declanleeler committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add understanding of dates for tasks

declanleeler committed:
Format code and bug fix

Fixed bug where commands were not being saved to the file despite changes to the list.
Reformatted code to adhere to code standards.

declanleeler committed:
Refactor code for OOP purposes

New classes have been added to extract different functionalities from Duke. They are mainly the
- Storage: deals with loading tasks from the file and saving tasks in the file
- TaskMaster: holds the list of tasks for the user, and has operations to edit the list
- Ui: deals with interactions with the user, focusing on taking in input and sending output
- Parser: deals with understanding the user's input, and reacts accordingly with the appropriate command
- Command: abstract class that deals with executing the appropriate actions to handle the input

The Command class has the following children:
- AddCommand: deals with creating and adds tasks to the taskMaster and storage.
- ExitCommand: ends the program and exits
- DeleteCommand: deletes the given task from taskMaster and saves to storage
- ListCommand: lists out the tasks in taskMaster
- MarkCommand: marks and unmarks tasks in taskMaster

declanleeler committed:
Organize classes into packages

declanleeler committed:
Add JUnit Tests

Add tests to test the behavior of the code for Parser and Ui.

declanleeler committed:
Add JavaDoc comments

Header comments added to all non-private classes/methods

declanleeler committed:
Tweak code to comply with CS2103 coding standard

declanleeler committed:
Add functionality to allow users to find tasks

Given a keyword, Duke can locate the tasks that contain this keyword in their description.

declanleeler committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak code to comply with CS2103 coding standard

# Conflicts:
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java

declanleeler committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add functionality to allow users to find tasks

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/task/Task.java

declanleeler committed:
Improve code quality

- Add comments for those that were missed out on the previous commit
- Edit naming of some variables to adhere to code standard
- Fixed minor bugs

zzkzzzz commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+            } myReader.close();
+        } catch (FileNotFoundException e) {
+            System.out.println("It seems you do not have an existing list, I will create it now.");
+            try {
+                File file = new File(path.toString());
+                file.getParentFile().mkdir();
+                file.createNewFile();
+                System.out.println("file created successfully");
+            } catch (IOException ioe) {
+                System.out.println("Failed to create file");
+            }
+        }
```
Should this be extracted out?

zzkzzzz commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,77 @@
+     * @throws DukeException
+     */
+    public AddCommand(String cmd, String details) throws DukeException {
+        if (details.length() == 0) {
+            throw new DukeException("☹ OOPS!!! The description cannot be empty.");
+        }
+        if (!details.contains("/")) {
+            throw new DukeException("☹ OOPS!!! The date of a " + cmd + " cannot be empty. " +
+                    "Use /at and type the date with the format yyyy-mm-dd after it");
+        }
```
Should this be extracted out to a method like `checkUserCommnand`?

declanleeler replied:
Yes that makes sense!

zzkzzzz commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,103 @@
+     * @param t task that was marked
+     * @param isComplete boolean flag for wehether the task was marked complete or not complete.
+     */
+    public void notifyMarkedTaskMessage(Task t, boolean isComplete) {
+        if (isComplete) {
+            System.out.println("Task " + t + " has been marked complete.");
+        } else {
+            System.out.println("Task " + t + " has been marked incomplete.");
+        }
+        printDoubleLine();
+    }
+
```
I like how you name the method name and variable name.

zzkzzzz commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    public static Command parse(String cmd, String details) throws DukeException {
+        switch (cmd) {
+        case "bye":
+            return new ExitCommand();
+        case "list":
+            return new ListCommand();
+        case "mark":
+            return new MarkCommand(details, true);
+        case "unmark":
+            return new MarkCommand(details, false);
+        case "delete":
+            return new DeleteCommand(details);
+        case "todo":
+            return new AddCommand(details);
+        case "deadline":
+        case "event":
+            return new AddCommand(cmd, details);
+        case "find":
+            return new FindCommand(details);
```
Perhaps change these command strings to constant variables?

zzkzzzz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+     * Constructor for Duke class.
+     * @param filePath the location of a textfile containing the user's list of tasks, relative to the root directory.
+     */
+    public Duke(String filePath) {
+        this.ui = new Ui();
+        this.storage = new Storage(filePath);
+        this.taskMaster = new TaskMaster(storage.loadTasks());
+    }
+
+    /**
+     * Kickstarts the Duke bot to begin taking in and reacting to user input.
+     */
+    public void run() {
+        ui.greet();
+        boolean isExit = false;
+        while (!isExit) {
+            try {
+                String[] userCommand = ui.readCommand();
+                Command command = Parser.parse(userCommand[0], userCommand[1]);
+                command.execute(this.taskMaster, this.ui, this.storage);
+                isExit = command.isExit();
+            } catch (DukeException e) {
+                System.out.println(e);
+            }
+        }
+    }
```
I like how you extract the code from the main function which increases the readability.

zzkzzzz reviewed (commented):
Overall, I found your code is very clear and easy to read. And almost all your code follows the Java code standard. I noted a few minor coding quality violations. Some of the methods are getting quite long. It would be better some of the code can be extracted.

kashing555 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,45 @@
+            return new FindCommand(details);
+        default:
+            throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
+    }
+
+
+}
```
Additional lines that can be removed

declanleeler replied:
Will delete it, thanks for checking!

kashing555 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+                String[] data = line.split("\\s*\\|\\s*");
+                String task = data[0];
+                switch (task) {
+                case "T":
+                    ToDo td = new ToDo(data[2]);
+                    if (data[1].equals("1")) {
+                        td.mark();
+                    }
+                    tasks.add(td);
+                    break;
+                case "D":
+                    Deadline deadline = new Deadline(data[2], LocalDate.parse(data[3], formatter));
+                    tasks.add(deadline);
+                    if (data[1].equals('1')) {
+                        deadline.mark();
+                    }
+                    break;
+                case "E":
+                    Event event = new Event(data[2], LocalDate.parse(data[3], formatter));
+                    tasks.add(event);
+                    if (data[1].equals('1')) {
+                        event.mark();
+                    }
+                    break;
+                }
```
Might want to tab the switch cases for better readability.

kashing555 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Kickstarts the actions to execute in order to carry out the command.
+     *
+     * @param tasks   holds all the tasks that the user has recorded down.
+     * @param ui      used to notify the user of task completion.
```
Unnecessary indentation for the description of parameters.

kashing555 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,50 @@
+import duke.exception.DukeException;
+import duke.task.Task;
+
+/**
+ *
+ */
```
Might want to add a description for class DeleteCommand. Empty comment seems out of place.

declanleeler replied:
Will add it in

kashing555 reviewed (commented):
Looks very good. Couldn't find any glaring errors. However, additional empty lines can be removed and some descriptions of methods/classes do not begin with a capital letter.

kashing555 commented on src/test/java/UiTest.java:
```diff
@@ -0,0 +1,43 @@
+
+public class UiTest {
+
+    private final PrintStream standardOut = System.out;
+    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
```
Final keywords should be all caps.

declanleeler replied:
Good catch, thanks!

declanleeler committed:
Remove extra lines in code

declanleeler committed:
Merge branch 'add-gradle-support' of https://github.com/declanleeler/ip

* 'add-gradle-support' of https://github.com/declanleeler/ip:
  Add Gradle support

declanleeler committed:
Automate project builds using Gradle

declanleeler committed:
Use CheckStyle

Use checkStyle to detect coding style violations and fix some of them.

declanleeler committed:
Add a GUI to Duke

Let's use JavaFX to implement the GUI so that users need not interact using the command line.

declanleeler committed:
Touch up GUI layout

declanleeler committed:
Add Assertions

Let's use assert feature to document important assumptions that should hold at various points in the code.
Assertions will help to
* indicate possible bugs
* terminate the program early when a bug is detected

declanleeler committed:
Improve code quality

Examine the code for flaws and improve its readability.

Let's,
- remove deep nesting
- add code comments
- ensure all names are in camel case
- remove all blank lines and redundant whitespaces

declanleeler committed:
Merge pull request #1 from declanleeler/branch-A-Assertions

Add Assertions

declanleeler committed:
Merge pull request #2 from declanleeler/branch-A-CodeQuality

Improve code quality

declanleeler committed:
Add archive functionality

There is no way to tidy Duke's list unless the user deletes tasks individually. These deleted tasks are not stored.

Let's allow users to archive completed tasks so that they can:
- retain a tidy list of uncompleted tasks in a single command
- do not need to delete completed tasks 1 by 1
- can store completed tasks in a text file when archived

Only completed tasks are removed and sent to the archive as incomplete tasks should remain on the to-do list unless specifically deleted.

declanleeler committed:
Coding standard quality improvements

declanleeler committed:
Improve naming conventions

declanleeler committed:
Improve messages to user

declanleeler committed:
Improve the GUI

- resize the icons to take less space
- bold the text for nicer readability
- remove truncation of text in dialog box

declanleeler committed:
Add screenshot of app to docs

declanleeler committed:
Add user guide
